staging/src/k8s.io/client-go/tools/pager/pager.go

```diff
@@ -115,3 +127,86 @@ func (p *ListPager) List(ctx context.Context, options metav1.ListOptions) (runti
 		options.Continue = m.GetContinue()
 	}
 }
+
+// EachListItem invokes fn on each runtime.Object in the list. Any error immediately terminates the
+// loop. The items are retrieved via List in small chunks from the server to reduce the impact on
+// the server. Up to ListPager.PageBufferSize chunks are buffered concurrently in the background.
+// If the chunk attempt fails a "Expired" error may be returned.
+func (p *ListPager) EachListItem(ctx context.Context, options metav1.ListOptions, fn func(obj runtime.Object) error) error {
+	return p.eachListChunkBuffered(ctx, options, func(obj runtime.Object) error {
+		return meta.EachListItem(obj, fn)
+	})
+}
+
+// stoppedErr is an internal error for signaling eachListChunk should stop processing chunks.
+// It must be handled internally and never be returned to API callers.
+var stoppedErr = goerrors.New("paginated list stopped")
+
+// eachListChunkBuffered invokes fn on each runtimeObject list chunk. It buffers up to
+// PageBufferSize chunks of list items concurrently in the background. If the chunk attempt fails a "Expired"
+// error may be returned.
+func (p *ListPager) eachListChunkBuffered(ctx context.Context, options metav1.ListOptions, fn func(obj runtime.Object) error) error {
+	chunkC := make(chan runtime.Object, p.PageBufferSize)
+
+	stopC := make(chan struct{})
+	bgResultC := make(chan error, 1)
+	go func() {
+		err := p.eachListChunk(ctx, options, func(chunk runtime.Object) error {
+			select {
+			case chunkC <- chunk: // buffer the chunk
+			case <-stopC:
+				return stoppedErr
+			}
+			return nil
+		})
+		if err == stoppedErr {
+			err = nil // stoppedErr is an internal signal
+		}
+		close(chunkC)
+		bgResultC <- err
+	}()
+
+	for o := range chunkC {
```

Comment:
If listing chunk in the background fails (such as due to resource expired), we will still finish processing the existing chunks in the buffer before return?

Reply:
Yes, I think from an API contract perspective, when a error occurs listing chunks, i think it's valid to either (1) stop calling fn as soon as possible, or (2) call fn on as many items as have been successfully retrieved.  I went with #1 since it was trivial to implement. I imagine in different situations a client might benefit more from one of these and in other situations the other. Thoughts?